### PR TITLE
Added basic filter functionality with regex search in stack.ParseYAML

### PR DESCRIPTION
**Pull request by ems5311**
Added basic filter functionality with regex search in stack.ParseYAML

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

## Description

This change adds filter functionality by introducing a new Persistent Flag which is called `--filter`. You can provide a regexp for this filter, and it will select functions from your YAML file (specified with `-f`) that are a regex match.

This works by removing everything from the parse YAML map `Functions` which is a member of the `stack.Services` struct.

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md))
#60 

## How Has This Been Tested?

Manually tested `faas-cli build` and `faas-cli deploy` with the `--filter` flag for various options.
EDIT: Tests now added

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -5,3 +5,4 @@ docker build --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_p
  docker cp faas-cli:/root/faas-cli . && \
  docker rm -f faas-cli
 
+go test ./...
```
Comment (alexellis) [inline build.sh]: What does this do? `./...`

File: stack/stack_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		title:        "Filter for all functions: '.*'",
+		filter:       ".*",
+		functions:    []string{"nodejs-echo", "ruby-echo", "shrink-image", "url-ping"},
+	},
+}
+
+func Test_parseYaml(t *testing.T) {
+	for _, test := range ParseYamlTests {
+		t.Run(test.title, func(t *testing.T) {
+			parsedYaml, err := ParseYAML("../samples.yml", test.filter)
```
Comment (alexellis) [inline stack/stack_test.go]: Brilliant to see the unit tests 💯 So that this test doesn't become brittle - let's create a test .yml file and keep it alongside the Golang code. Even better if you add an override which takes []bytes or string rather than a filename - then we can hard-code the test case into the file and use multiple inputs.

File: build.sh (hunk 1 above)
Comment (ems5311) [reply build.sh]: Runs all unit tests in the entire project, directory by directory. If it encounters a directory without tests, it just does nothing. It the dir has tests, it runs them.

File: stack/stack_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	{
+		title:        "Filter functions only containing '.+-.+'",
+		filter:       ".+-.+",
+		functions:    []string{"nodejs-echo", "ruby-echo", "url-ping"},
+	},
+	{
+		title:        "Filter for all functions: '.*'",
+		filter:       ".*",
+		functions:    []string{"imagemagick", "nodejs-echo", "ruby-echo", "url-ping"},
+	},
```
Comment (johnmccabe) [inline stack/stack_test.go]: You should add some negative tests, for example what happens when the filter results in an empty function slice.

File: stack/stack.go [hunk 4]
```diff
@@ -43,11 +50,21 @@ func ParseYAML(yamlFile string) (*Services, error) {
...
+		for k, function := range services.Functions {
+			function.Name = k
+			match, _ := regexp.MatchString(filter, function.Name)
+			if !match {
+				delete(services.Functions, function.Name)
+			}
+		}
+	}
+
 	return &services, err
```
Comment (johnmccabe) [inline stack/stack.go]: Shouldn't the error be `nil` here?

File: stack/stack.go [hunk 5]
```diff
@@ -32,6 +32,13 @@ func ParseYAML(yamlFile string) (*Services, error) {
...
 	}
+	return ParseYamlData(fileData, filter)
+}
+
+// ParseYamlData parse YAML data into a stack of "services".
+func ParseYamlData(fileData []byte, filter string) (*Services, error) {
+	var services Services
+	var err error
 
 	err = yaml.Unmarshal(fileData, &services)
```
Comment (johnmccabe) [inline stack/stack.go]: can just use a short assignment here `:=` and dump the `err` forward declaration on line 41.

File: stack/stack_test.go (hunk 3 above)
Comment (johnmccabe) [reply stack/stack_test.go]: Also what happens when the regex is invalid.

File: stack/stack.go [hunk 6]
```diff
@@ -43,11 +50,21 @@ func ParseYAML(yamlFile string) (*Services, error) {
 		return nil, fmt.Errorf("'%s' is the only valid provider for this tool - found: %s", providerName, services.Provider.Name)
 	}
 
+	if len(filter) > 0 {
+		for k, function := range services.Functions {
+			function.Name = k
+			match, _ := regexp.MatchString(filter, function.Name)
```
Comment (johnmccabe) [inline stack/stack.go]: should handle the error here and exit with a message warning the user why the command has failed

File: stack/stack.go (hunk 4 above)
Comment (ems5311) [reply stack/stack.go]: Do you mean I should set the `err` to `nil` explicitly before returning it?

Comment (johnmccabe) [reply stack/stack.go]: for clarity yes

File: stack/stack_test.go (hunk 3 above)
Comment (ems5311) [reply stack/stack_test.go]: When the function results in no functions being matched, the same behavior will be exhibited as if you had called `faas-cli <verb> -f <YAML>` on a YAML file which does not have a `functions:` object. What it does with the `build` action is prints the following error message.

`Please provide a valid -image name for your Docker image.`

I don't think it's within the scope of this PR to fix the behavior of commands when they parse a YAML file and there are no functions.

If the regex is invalid, the err variable will be set (I will change it to storing and checking the `err` variable after the call to `regexp.MatchString()`)

Comment (ems5311) [reply stack/stack_test.go]: Actually, I was wrong about the error behavior. I found a case that causes the `err` var to be set to non-`nil`:

```
faas-cli deploy -f samples.yml --filter "*"
2017/09/06 06:32:49 Error parsing regex passed in through --filter argument: error parsing regexp: missing argument to repetition operator: `*`
2017/09/06 06:32:49 error parsing regexp: missing argument to repetition operator: `*`
```

If you issue a command with quotes around the `*` variable without an argument, the go `regexp` parser errors out. I think this is another reason why filter should be wildcard matching rather than regexp matching. I think we can use [this class for wildcard matching](https://github.com/ryanuber/go-glob/blob/master/glob.go).

**Comment by ems5311** [comment]
@johnmccabe Addressed your comments as best I could in the latest commit. Added a few test cases.

File: stack/stack.go [hunk 7]
```diff
@@ -43,11 +49,26 @@ func ParseYAML(yamlFile string) (*Services, error) {
...
+				fmt.Printf("Error parsing regex passed in through --filter argument: %v", err)
+				return nil, err
+			}
+			if !match {
+				delete(services.Functions, function.Name)
+			}
+		}
+	}
+
+	err = nil
```
Comment (johnmccabe) [inline stack/stack.go]: Better to just return nil, don't bother assigning a value to err
```
return &services, nil
```

File: stack/stack.go [hunk 8]
```diff
@@ -43,11 +49,26 @@ func ParseYAML(yamlFile string) (*Services, error) {
 		return nil, fmt.Errorf("'%s' is the only valid provider for this tool - found: %s", providerName, services.Provider.Name)
 	}
 
+	if len(filter) > 0 {
+		for k, function := range services.Functions {
+			function.Name = k
+			match, err := regexp.MatchString(filter, function.Name)
+			if err != nil {
+				fmt.Printf("Error parsing regex passed in through --filter argument: %v", err)
```
Comment (johnmccabe) [inline stack/stack.go]: Minor tweak I'd change the error to `Invalid regex passed to filter`

File: stack/stack.go [hunk 9]
```diff
@@ -43,11 +49,26 @@ func ParseYAML(yamlFile string) (*Services, error) {
...
+	}
+
+	err = nil
 	return &services, err
 }
 
-// fetchYAML pulls in file from remote location such as GitHub raw file-view
-func fetchYAML(address *url.URL) ([]byte, error) {
+// fetchYaml pulls in file from remote location such as GitHub raw file-view
+func fetchYaml(address *url.URL) ([]byte, error) {
```
Comment (johnmccabe) [inline stack/stack.go]: should be fetchYAML (in comments and fn calls etc)

File: stack/stack_test.go (hunk 3 above)
Comment (johnmccabe) [reply stack/stack_test.go]: @ems5311  we've had a chat about this, and think your glob suggestion is a good one, we'd still like to keep the regex handling however so how about two flags:
- `--filter=` using your glob suggestion for a simple filtering experience
- `--regex=` using the current regex handling that you have here for a richer filtering

Thanks for all your work on this btw, its looking great so far !!!

File: stack/stack.go (hunk 9 above)
Comment (ems5311) [reply stack/stack.go]: blah, missed this one. Good catch

**Comment by johnmccabe** [comment]
Can you also add an example to each of the command declarations, for example in `commands/deploy.go`, in the `deployCmd` var:
```
	Example: `  faas-cli deploy -f https://domain/path/myfunctions.yml
  faas-cli deploy -f ./samples.yml
  faas-cli deploy -f ./samples.yml --replace=false
  faas-cli deploy -f ./samples.yml --filter="*gif*"
  faas-cli deploy -f ./samples.yml --regex="fn[0-9]_.*"
```

**Comment by alexellis** [comment]
Have you covered all of John's feedback?

**Comment by ems5311** [comment]
Now I have addressed all of John's feedback, also added checks for if `--filter` and `--regex` is passed in, and added a test for that. Added some negative tests for `--regex`. `--filter` does not return any `error`, so I think the current test coverage is OK.

I have squashed these changes into one commit.

**Comment by alexellis** [comment]
That sounds perfect, thank you so much! 

I'll push a dev-build, if you can get 1-2 people to test it on the Slack channel I'll merge.

**Comment by alexellis** [comment]
https://github.com/alexellis/faas-cli/releases/tag/0.4.13-alpha

**Comment by alexellis** [comment]
Thank you @ems5311 - I'll merge and we'll need to test this extensively in the next pre-release build.
